openlibrary/core/observations.py

```diff
@@ -1,26 +1,245 @@
 """Module for handling patron observation functionality"""
 
+import json
 import requests
 
 from infogami import config
 from openlibrary import accounts
+
 from . import cache
+from . import db
+
+
+@cache.memoize(engine="memcache", key="observations", expires=config.get('observation_cache_duration'))
+def get_observations():
+    """
+    Returns a dictionary of observations that are used to populate forms for patron feedback about a book.
+
+    Dictionary has the following structure:
+    {
+        'observations': [
+            {
+                'id': observation_types.id,
+                'label': observation_types.type,
+                'description': observation_types.description,
+                'multi_choice': observation_types.allow_multiple_values,
+                'values': [ observation_values.value list ]
+            }
+        ]
+    }
+
+    return: Dictionary of all possible observations that can be made about a book.
+    """
+    observations = Observations.get_observation_types_and_values()
+
+    observation_dict = {}
+
+    for o in observations:
+        type = o['type']
+
+        if type not in observation_dict:
+            observation_dict[type] = {
+                'id': o['type_id'],
+                'label': type,
+                'description': o['description'],
+                'multi_choice': o['multi_choice'],
+                'values': []
+            }
+
+        observation_dict[type]['values'].insert(0, (o['value_id'], o['value'], o['prev_value']))
+
+    observation_list = []
 
-# URL for TheBestBookOn
-TBBO_URL = config.get('tbbo_url')
+    for v in observation_dict.values():
+        v['values'] = _sort_values(v['values'])
+        observation_list.append(v)
 
-def post_observation(data, s3_keys):
-    headers = {
-        'x-s3-access': s3_keys['access'],
-        'x-s3-secret': s3_keys['secret']
+    response = {
+        'observations': observation_list
     }
 
-    response = requests.post(TBBO_URL + '/api/observations', data=data, headers=headers)
+    return response
+
+def _sort_values(values_list):
+    """
+    Given a list of value tuples, returns a sorted list of observation values.
+    
+    Value tuples have the form (value ID, value, previous value ID).  Previous
+    value ID is the ID of the previous value in the list, or None if the value
+    is the first item in the list.
+
+    return: A sorted list of values.
+    """
+    # Add middle list item to sorted list
+    middle_item = values_list.pop(int(len(values_list) / 2))
+    sorted_list = [ middle_item[1] ]
+
+    # Previous id:
+    prev = middle_item[2]
+
+    # Next id:
+    next = middle_item[0]
+
+    map = {}
```

Comment:
This could be a dict comprehension.

Would it be possible to find a variable that does not shadow the Python builtin https://docs.python.org/3/library/functions.html#ma ?